src/main/java/com/monumental/services/MonumentService.java

```diff
@@ -157,8 +159,41 @@ public Integer countSearchResults(String searchQuery, String latitude, String lo
         Root<Monument> root = query.from(Monument.class);
         query.select(builder.countDistinct(root));
 
-        this.buildSearchQuery(builder, query, root, searchQuery, latitude, longitude, distance);
+        this.buildSearchQuery(builder, query, root, searchQuery, latitude, longitude, distance, false);
 
         return this.getEntityManager().createQuery(query).getSingleResult().intValue();
     }
+
+    /**
+     * Gets the related monuments and sets them on the monument objects, using only one extra SQL query
+     * @param monuments Monuments to get tags for - these objects are updated directly using setTags
+     *                  but no database update is called
+     */
+    public void getRelatedTags(List<Monument> monuments) {
+        CriteriaBuilder builder = this.getCriteriaBuilder();
+        CriteriaQuery<Monument> query = this.createCriteriaQuery(builder, false);
+        Root<Monument> root = this.createRoot(query);
+        query.select(root);
+        root.fetch("tags", JoinType.LEFT);
+
+        List<Integer> ids = new ArrayList<>();
+        for (Monument monument : monuments) {
+            ids.add(monument.getId());
+        }
+
+        query.where(
+                root.get("id").in(ids)
+        );
+
+        List<Monument> monumentsWithTags = this.getWithCriteriaQuery(query);
+
+        Map<Integer, List<Tag>> tagsMap = new HashMap<>();
+        for (Monument monument : monumentsWithTags) {
+            tagsMap.put(monument.getId(), monument.getTags());
+        }
+
+        for (Monument monument : monuments) {
+            monument.setTags(tagsMap.get(monument.getId()));
+        }
+    }
```

Comment:
What's the point of using `tagsMap` and `setTags`?

Won't `monumentsWithTags` return all of the `Monument`s with their `Tag`s already?

Reply:
Yes but they're in the wrong order and not paginated.